### PR TITLE
Task 3

### PR DESCRIPTION
**Pull request by anastasiyakoush**

### CODE REVIEW

File: react/counter-app/src/App.js [hunk 1]
```diff
@@ -1,8 +1,29 @@
...
+import FormContainer from "./containers/FormContainer";
+import NoMatch from "./components/NoMatch/index";
+import LoginRedux from "./containers/LoginRedux";
+import LoginSuccess from "./components/LoginSuccess/index";
 import ParentContainer from "./containers/ParentContainer";
 
 const App = () => {
-  return <ParentContainer />;
+  return (
+    <BrowserRouter>
```
Comment (mxbelski) [inline react/counter-app/src/App.js]: 1. Move router to separate file. Import it into App.js and use.
2. Move route values into  const object. Reuse it in router.

File: react/counter-app/src/App.js [hunk 2]
```diff
@@ -1,8 +1,29 @@
...
+      <NavBar />
+      <Switch>
+        <Route exact path="/" />
+        <Route path="/about" component={About} />
+        <Route path="/counters" component={ParentContainer} />
+        <Route path="/login" component={FormContainer} />
+        <Route exact path="/login-redux" component={LoginRedux} />
+        <Route path="/login-redux/success" component={LoginSuccess} />
+        <Route path="/404" component={NoMatch} />
```
Comment (mxbelski) [inline react/counter-app/src/App.js]: Use "NotFound" instead of "NoMatch"

File: react/counter-app/src/components/About/index.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+import { aboutCounters } from "../../consts";
+import { Container, Paper } from "@material-ui/core";
+import { withStyles } from "@material-ui/styles";
+import styles from "./styles";
+
+const About = props => {
+  const { classes } = props;
```
Comment (mxbelski) [inline react/counter-app/src/components/About/index.js]: const { classNames }

File: react/counter-app/src/components/About/index.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+  return (
+    <Container>
+      <Paper className={classes.container}>{aboutCounters}</Paper>
+    </Container>
+  );
+};
+
+export default withStyles(styles)(About);
```
Comment (mxbelski) [inline react/counter-app/src/components/About/index.js]: good. I like it.

File: react/counter-app/src/components/Form/index.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        value={props.email}
+        name="email"
+        autoComplete="false"
+        onChange={props.handleEmailChange}
+        onKeyPress={props.onKeyPressHandler}
+        className={classes.input}
+        required
+      />
+
+      {!props.isEmailValid && tipField(`Please use correct email`)}
```
Comment (mxbelski) [inline react/counter-app/src/components/Form/index.js]: move validation errors before input field.

File: react/counter-app/src/components/Form/index.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        value={props.password}
+        name="password"
+        onChange={props.handlePasswordChange}
+        onKeyPress={props.onKeyPressHandler}
+        className={classes.input}
+        required
+      />
+
+      {!props.isPasswordValid &&
+        tipField(`Password must be at least ${props.passwordMinLength}`)}
```
Comment (mxbelski) [inline react/counter-app/src/components/Form/index.js]: same here. Before field.

File: react/counter-app/src/components/Form/index.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+import { withStyles } from "@material-ui/styles";
+import styles from "./styles";
+
+const LoginForm = props => {
```
Comment (mxbelski) [inline react/counter-app/src/components/Form/index.js]: You must not redirect to success screen with invalid username/password combinations. You must stay at login page and warn user until he changes his credentials to valid.

File: react/counter-app/src/components/Form/index.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        type="email"
+        value={props.email}
+        name="email"
+        autoComplete="false"
+        onChange={props.handleEmailChange}
+        onKeyPress={props.onKeyPressHandler}
+        className={classes.input}
+        required
+      />
+
```
Comment (mxbelski) [inline react/counter-app/src/components/Form/index.js]: remove empty spaces in markup sections.

File: react/counter-app/src/components/InputReflector/index.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { withStyles } from "@material-ui/styles";
+import styles from "./styles";
+
+const InputReflector = props => {
```
Comment (mxbelski) [inline react/counter-app/src/components/InputReflector/index.js]: I think, InputDisplay is enough :)

File: react/counter-app/src/containers/FormContainer.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  const submitHadler = () => {
+    if (email !== "" && password !== "" && isEmailValid && isPasswordValid) {
+      console.log(email);
+      console.log(password);
+      setEmail({ email: "", isEmailValid: true });
+      setPassword({ password: "", isPasswordValid: true });
+    } else console.log("error");
+  };
+  const onKeyPressHandler = event => {
+    event.key === "Enter" && submitHadler();
```
Comment (mxbelski) [inline react/counter-app/src/containers/FormContainer.js]: Its better to use key code here

File: react/counter-app/src/containers/FormContainer.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    });
+  };
+
+  const submitHadler = () => {
+    if (email !== "" && password !== "" && isEmailValid && isPasswordValid) {
+      console.log(email);
+      console.log(password);
+      setEmail({ email: "", isEmailValid: true });
+      setPassword({ password: "", isPasswordValid: true });
+    } else console.log("error");
```
Comment (mxbelski) [inline react/counter-app/src/containers/FormContainer.js]: else {
    /*do something*/
}

File: react/counter-app/src/containers/FormContainer.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+  const handlePasswordChange = event => {
+    setPassword({
+      password: event.target.value,
+      isPasswordValid: validateLength(password, PASSWORD_MIN_LENGTH)
+    });
+  };
+
+  const submitHadler = () => {
+    if (email !== "" && password !== "" && isEmailValid && isPasswordValid) {
```
Comment (mxbelski) [inline react/counter-app/src/containers/FormContainer.js]: Are you sure you cannot use here just if (email && password && ... etc)

File: react/counter-app/src/containers/FormContainer.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import React, { useState } from "react";
+import LoginForm from "../components/Form/index";
+import InputReflector from "../components/InputReflector/index";
+import { PASSWORD_MIN_LENGTH, emailRegEx } from "../consts";
+import { validateLength } from "../validators";
+
+const FormContainer = props => {
```
Comment (mxbelski) [inline react/counter-app/src/containers/FormContainer.js]: redundant props

File: react/counter-app/src/containers/FormContainer.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  };
+
+  const handlePasswordChange = event => {
+    setPassword({
+      password: event.target.value,
+      isPasswordValid: validateLength(password, PASSWORD_MIN_LENGTH)
+    });
+  };
+
+  const submitHadler = () => {
```
Comment (mxbelski) [inline react/counter-app/src/containers/FormContainer.js]: Handler

File: react/counter-app/src/containers/LoginRedux.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    if (isEmailValid && isPasswordValid) {
+      props.login(email, password);
+      props.history.push("login-redux/success");
+    } else console.log("errrrr");
+  };
+
+  const onKeyPressHandler = event => {
+    event.key === "Enter" && submitHadler();
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: keycode

File: react/counter-app/src/containers/LoginRedux.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        isEmailValid={isEmailValid}
+        isPasswordValid={isPasswordValid}
+        buttonText={isAuthenticated ? "Log out" : "Log in"}
+      />
+      <InputReflector email={email} password={password} />
+    </>
+  );
+};
+
+const mapStateToProps = state => ({ ...state });
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: Same comment related to this mapStateToProps representation as one of my comments above.

File: react/counter-app/src/containers/LoginRedux.js [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+const mapDispatchToProps = dispatch => {
+  return {
+    setEmail: value => dispatch(actions.setEmail(value)),
+
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: remove empty space

File: react/counter-app/src/containers/LoginRedux.js [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    setPassword: value => dispatch(actions.setPassword(value)),
+
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: remove empty space

File: react/counter-app/src/containers/LoginRedux.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    validateEmail: constraint => dispatch(actions.validateEmail(constraint)),
+
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: remove empty space

File: react/counter-app/src/containers/LoginRedux.js [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    validatePassword: constraint =>
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: Write in one lina and remove empty space

File: react/counter-app/src/containers/LoginRedux.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+      dispatch(actions.validatePassword(constraint)),
+
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: remove empty space

File: react/counter-app/src/containers/LoginRedux.js [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    login: (email, password) => dispatch(actions.login(email, password)),
+
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: remove empty space

File: react/counter-app/src/components/About/index.js (hunk 3 above)
Comment (anastasiyakoush) [reply react/counter-app/src/components/About/index.js]: const { classes: classNames } = props;
можно так, иначе withStyles работать не будут

File: react/counter-app/src/components/About/index.js [hunk 23]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { aboutCounters } from "../../consts";
+import { Container, Paper } from "@material-ui/core";
+import { withStyles } from "@material-ui/styles";
+import styles from "./styles";
+
+const About = props => {
+    const { classes: classNames } = props;
+
+
```
Comment (mxbelski) [inline react/counter-app/src/components/About/index.js]: Double /n. Please, remove one of this.

File: react/counter-app/src/components/AppRouter.js [hunk 24]
```diff
@@ -0,0 +1,28 @@
+import React from "react";
+import { Route, Switch, Redirect, withRouter } from "react-router-dom";
+import { routes, root } from "../consts";
+import About from "./About/index";
+import FormContainer from "../containers/FormContainer";
+import NotFound from "./NotFound/index";
```
Comment (mxbelski) [inline react/counter-app/src/components/AppRouter.js]: Import order is missing.

File: react/counter-app/src/containers/FormContainer.js [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+    });
+
+    const [{ password, isPasswordValid }, setPassword] = useState({
+        password: "",
+        isPasswordValid: true
+    });
+
+    const handleEmailChange = event => {
+        const email = event.target.value;
+        const isEmailValid = validateEmail(email, emailRegEx);
```
Comment (mxbelski) [inline react/counter-app/src/containers/FormContainer.js]: isValid is enough.
Same for passowrd.

File: react/counter-app/src/containers/LoginRedux.js [hunk 26]
```diff
@@ -0,0 +1,103 @@
+import React, { Component } from "react";
+import { connect } from "react-redux";
+import { withRouter } from "react-router-dom";
+import * as actions from "../redux/actions";
+import LoginForm from "../components/Form/index";
+import InputDisplay from "../components/InputDisplay/index";
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: import order is missing.

File: react/counter-app/src/containers/LoginRedux.js [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+import { PASSWORD_MIN_LENGTH, emailRegEx, routes, root } from "../consts";
+import { validatePassword, validateEmail } from "../validators";
+
+class LoginRedux extends Component {
+    constructor(props) {
+        super(props);
+        this.state = { ...props };
+        this.state.ok = false;
```
Comment (mxbelski) [inline react/counter-app/src/containers/LoginRedux.js]: Ok is not valid name.

**Pull request merged**